Nested state machine2

Eyap53 opened the pull request:
**Generic PR**  
- Do you have a forum thread linked to this PR?  
Yes : [threads/state-machine.979371](https://forum.unity.com/threads/state-machine.979371/page-7)
- What will this PR bring to the project for everyone?  
It had a new nice functionality (nested StateMachine), that also help with separating the StateMachine, making things more comprehensive when looking at the Table. See below for an example.

- How did you implement them?  
I just added a new ActionSO, that reference the SM-table.

Practical example : 

As it may rise UX questions, I decided to implement a little example, which is also in this PR. It concerns the StateMachine table of the slime critter.

The StateMachine for the slime critter was like this : 
![initial-table](https://user-images.githubusercontent.com/24421318/106942007-89e20100-671b-11eb-87f2-1d684b9bc1d0.png)


and is now reduced to this :
![end_table](https://user-images.githubusercontent.com/24421318/106942052-98c8b380-671b-11eb-9890-f4be6aed6f53.png)

It's still complicated, but a people taking a look at this table won't have to understand the relations between the two previous states. Instead, when they understand the fighting relations, they can then go look at the roaming table : 
![roaming_table](https://user-images.githubusercontent.com/24421318/106942103-ad0cb080-671b-11eb-9556-795a5bf759bb.png)

In the process, a few transitions were also removed. Indeed, previously both Idle and the previous Roaming were adding some condition, that are not necessary anymore : 

![not_needed2](https://user-images.githubusercontent.com/24421318/106944274-49d04d80-671e-11eb-91d1-65a3a37be453.png)

Note that this other table can be used as-is for another critter, even if the parent table is different ! So it also add re-usability.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/UnityTechnologies/open-project-1?pullRequest=353) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/UnityTechnologies/open-project-1?pullRequest=353) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/UnityTechnologies/open-project-1?pullRequest=353) it.</sub>

ciro-unity commented:
Hmm, having the "example" build on top of existing transition tables means we will probably have a conflict once I'm done with the PRs I'm reviewing... isn't it possible to build it as an independent one?

Eyap53 commented:
I restored the previous SM table, and created a new one with the mentioned improvements. I switched the SM table reference in the Slime Critter prefab to the new one (the simplified version).
I also reflected the improvements made in the last PRs to the simplified table.

Note that while adding the new condition "IsTargetDead" to the old SM, the transition from the "Roaming" state to the "Idle" state was forgotten (Do not cause a big visual problem, but I don't think it was intentional). Thus proving my previous point that redundancy is error-prone and the validity of my example. I fixed this transition in my last commit.
